lib/api/v3/work_packages/schema/work_package_schema.rb

```diff
@@ -77,6 +77,22 @@ def assignable_categories
           def available_custom_fields
             @project.all_work_package_custom_fields & @type.custom_fields.all
           end
+
+          def percentage_done_writable?
+            if Setting.work_package_done_ratio == 'status' ||
+               Setting.work_package_done_ratio == 'disabled'
+              return false
+            end
+            leaf_or_nil?(@work_package)
+          end
+
+          def estimated_time_writable?
+            leaf_or_nil?(@work_package)
+          end
+
+          def leaf_or_nil?(work_package)
+            work_package.nil? ? true : work_package.children.empty?
```

Comment:
@NobodysNightmare you are right, `leaf?` is the fastest option as it does not require any DB interaction.